### PR TITLE
Add support for CMSIS Neural Network

### PR DESCRIPTION
**Pull request by jerpelea**
## Summary
the CMSIS NN software library is a collection of efficient neural
network kernels developed to maximize the performance and minimize
the memory footprint of neural networks on Cortex-M processor cores.

Project https://github.com/ARM-software/CMSIS_5

The library is divided into a number of functions each covering
a specific category:

    Convolution Functions
    Activation Functions
    Fully-connected Layer Functions
    SVDF Layer Functions
    Pooling Functions
    Softmax Functions
    Basic math Functions

The library has separate functions for operating on different weight
and activation data types including 8-bit integers (q7_t) and 16-bit
integers (q15_t). The descrition of the kernels are included in the
function description.

More information
https://www.keil.com/pack/doc/CMSIS/NN/html/index.html

Project license : Apache 2.0 License
https://github.com/ARM-software/CMSIS_5/blob/develop/LICENSE.txt

## Impact
NN

## Testing
Spresense

### CODE REVIEW

File: mlearning/libcmsisdsp/Kconfig [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+config CMSIS
```
[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Kconfig:6]: I preffer them splitted

File: mlearning/libcmsisdsp/Makefile [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mult_real_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_dot_prod_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_conj_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mult_real_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mag_f32.c
+
+CFLAGS += -fno-strict-aliasing -ffunction-sections -fdata-sections
+CFLAGS += -ffast-math -Wno-undef -Wno-shadow -Wno-undef
+
+BIN := libarm_cortexM4lf_math$(LIBEXT)
```
[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Makefile:556]: why should I remove the cmsis_dsp?

[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Makefile:556]: libarm_cortexM4lf_math is needed

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply mlearning/libcmsisdsp/Makefile:555]: CMSIS is a general library could be run on many arm architecture, it's wrong to add cortexM4lf in the library name.

[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Makefile:555]: Thanks for explaining what you consider wrong.

[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Makefile]: fixed

File: mlearning/libcmsisdsp/Kconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply mlearning/libcmsisdsp/Kconfig:6]: CMSIS contain both DSP and NN, it doesn't make sense to download CMSIS package to mlearning/libcmsisdsp, and reference it from mlearning/libcmsisnn. How about:
mlearning/cmsis/
CMSIS_5
libcmsisdsp
  Kconfig
  Make.defs
  Makefile
libcmsisnn
  Kconfig
  Make.defs
  Makefile
Kconfig
Make.def
Makefile

File: mlearning/libcmsisnn/Makefile [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+ifeq ($(CONFIG_CMSIS_NN),y)
+INCLUDES += $(TOPDIR)/include/cmsis/
```
Comment (xiaoxiang781216) [inline mlearning/libcmsisnn/Makefile:24]: where is include/cmsis?

File: mlearning/libcmsisnn/Makefile [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+ifeq ($(CONFIG_CMSIS_NN),y)
```
Comment (xiaoxiang781216) [inline mlearning/libcmsisnn/Makefile:23]: don't ifeq here

File: mlearning/libcmsisdsp/Makefile [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+ifeq ($(CONFIG_CMSIS_DSP),y)
```
Comment (xiaoxiang781216) [inline mlearning/libcmsisdsp/Makefile:23]: don't need check.

File: mlearning/libcmsisnn/Makefile (hunk 3 above)
Comment (jerpelea) [reply mlearning/libcmsisnn/Makefile:24]: leftover from initial patch

File: mlearning/libcmsisdsp/Makefile (hunk 5 above)
Comment (jerpelea) [reply mlearning/libcmsisdsp/Makefile]: thanks for the review

File: mlearning/libcmsisdsp/Makefile [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+ifeq ($(CONFIG_EXTERNALS_CMSIS_DSP_ARM_MATH_MATRIX_CHECK),y)
+CFLAGS += -DARM_MATH_MATRIX_CHECK
```
Comment (xiaoxiang781216) [inline mlearning/libcmsisdsp/Makefile:24]: The flag add here isn't enough to get the full optimized code:
https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/DSP/config.cmake
https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/DSP/cmsisdspconfig.py
You need add the same flag as the original cmake script.

File: mlearning/libcmsisdsp/Kconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerpelea) [reply mlearning/libcmsisdsp/Kconfig:6]: thanks for the suggestion. I'll do it

File: mlearning/cmsis/Makefile [hunk 7]
```diff
@@ -0,0 +1,34 @@
+############################################################################
+# apps/mlearning/libcmsisdsp/Makefile
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/Makefile:2]: libcmsisdsp->cmsis

File: mlearning/cmsis/libcmsisdsp/Make.defs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+#
+############################################################################
+
+ifeq ($(CONFIG_CMSIS_DSP),y)
+CONFIGURED_APPS += $(APPDIR)/mlearning/cmsis/libcmsisdsp
+
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/Include/}
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/NN/Include/}
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Make.defs:27]: move to libcmsisnn/Make.defs

File: mlearning/cmsis/libcmsisdsp/Make.defs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/Include/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/NN/Include/}
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Make.defs:31]: move to libcmsisnn/Make.defs

File: mlearning/cmsis/libcmsisnn/Make.defs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+ifeq ($(CONFIG_CMSIS_NN),y)
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisnn/Make.defs:21]: ifeq ($(CONFIG_CMSIS_NN),)

File: mlearning/cmsis/libcmsisdsp/Make.defs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+ifeq ($(CONFIG_CMSIS_DSP),y)
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Make.defs:21]: ifneq ($(CONFIG_CMSIS_DSP),)

File: mlearning/cmsis/Make.defs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(wildcard $(APPDIR)/mlearning/cmsis/*/Make.defs)
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/Make.defs:21]: ifneq ($(CONFIG_CMSIS),)
CONFIGURED_APPS += $(APPDIR)/mlearning/cmsis

include $(wildcard $(APPDIR)/mlearning/cmsis/*/Make.defs)
endif

File: mlearning/cmsis/libcmsisnn/Makefile [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_mat_mul_core_4x_s8.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_mat_mult_nt_t_s8.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_accumulate_q7_to_q15.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_depthwise_conv_nt_t_s8.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_mat_mul_core_1x_s8.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_nn_add_q7.c
+CSRCS += $(CMSIS5_NN)/NNSupportFunctions/arm_q7_to_q15_reordered_with_offset.c
+
+BIN := libcmsis_nn$(LIBEXT)
+VPATH += libcmsisnn
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisnn/Makefile:99]: don't need VPATH

File: mlearning/cmsis/libcmsisdsp/Makefile [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_dot_prod_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_conj_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mult_real_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mag_f32.c
+
+CFLAGS += -fno-strict-aliasing -ffunction-sections -fdata-sections
+CFLAGS += -ffast-math -Wno-undef -Wno-shadow -Wno-undef
+
+BIN := libarm_math$(LIBEXT)
+VPATH += :libcmsisdsp
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Makefile:554]: no needed VPATH

File: mlearning/cmsis/libcmsisdsp/Makefile [hunk 15]
```diff
@@ -0,0 +1,556 @@
...
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/ComplexMathFunctions.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mag_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/ComplexMathFunctionsF16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mult_real_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_dot_prod_f16.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_conj_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mult_real_q31.c
+CSRCS += $(CMSIS5_DSP)/ComplexMathFunctions/arm_cmplx_mag_f32.c
+
+CFLAGS += -fno-strict-aliasing -ffunction-sections -fdata-sections
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Makefile:550]: remove, set in board's Make.defs

File: mlearning/cmsis/libcmsisdsp/Makefile [hunk 16]
```diff
@@ -0,0 +1,556 @@
...
+CFLAGS += -ffast-math -Wno-undef -Wno-shadow -Wno-undef
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Makefile:551]: remove the last dup -Wno-undef, and should we define the macro to the known value and then remove -Wno-undef?

File: mlearning/cmsis/libcmsisnn/Makefile [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+CFLAGS += -Wno-undef
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisnn/Makefile:23]: should we define the macro to the known value

File: mlearning/cmsis/.gitignore [hunk 18]
```diff
@@ -0,0 +1 @@
+/*.zip
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/.gitignore:1]: add new line

File: mlearning/cmsis/libcmsisnn/Make.defs (hunk 10 above)
Comment (xiaoxiang781216) [reply mlearning/cmsis/libcmsisnn/Make.defs:21]: Not change?

File: mlearning/cmsis/libcmsisdsp/Make.defs [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+ifneq ($(CONFIG_CMSIS_DSP),)
+CONFIGURED_APPS += $(APPDIR)/mlearning/cmsis/libcmsisdsp
+
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/Include/}
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/Include/}
+CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
```
Comment (xiaoxiang781216) [inline mlearning/cmsis/libcmsisdsp/Make.defs:29]: should we move
```
CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/Core/Include/}
```
to apps/mlearning/cmsis/Make.defs and move:
```
CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
CXXFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(APPDIR)/mlearning/cmsis/CMSIS_5/CMSIS/DSP/PrivateInclude/}
```
to mlearning/cmsis/libcmsisdsp/Makefile?

File: mlearning/libcmsisdsp/Makefile (hunk 6 above)
Comment (xiaoxiang781216) [reply mlearning/libcmsisdsp/Makefile]: Ok.

**Comment by jerpelea** [comment]
@xiaoxiang781216 done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**